README.rst
###################################
Multi-threaded Optimization Toolbox
###################################
The Multi-threaded Optimization Toolbox (MOT) is a library for parallel optimization and sampling using the OpenCL compute platform.
Using OpenCL allows parallel processing using all CPU cores or using the GPU (Graphics card).
MOT implements OpenCL parallelized versions of the Powell, Nelder-Mead Simplex and Levenberg-Marquardt non-linear optimization algorithms
alongside various flavors of Markov Chain Monte Carlo (MCMC) sampling.

For the full documentation see: https://mot.readthedocs.org


****************
Can MOT help me?
****************
MOT can help you if you have multiple small independent optimization problems.
For example, if you have a lot of (>10.000) small optimization problems, with ~30 parameters or less each, MOT may be of help.
If, on the other hand, you have one big optimization problem with 10.000 variables, MOT unfortunately can not help you.


****************
Example use case
****************
MOT was originally written as a computation package for the `Microstructure Diffusion Toolbox <https://github.com/robbert-harms/MDT>`_, used in dMRI brain research.
In diffusion Magnetic Resonance Imaging (dMRI) the brain is scanned in a 3D grid where each grid element, a *voxel*, represents its own optimization problem.
The number of data points per voxel is generally small, ranging from 30 to 500 datapoints, and the models fitted to that data have generally
somewhere between 6 and 20 parameters.
Since each of these voxels can be analyzed independently of the others, the computations can be massively parallelized and hence programming
in OpenCL potentially allows large speed gains.
This software toolbox was originally built for exactly this use case, yet the algorithms and data structures are generalized such that any
scientific field may take advantage of this toolbox.

For the diffusion MRI package *MDT* to which is referred in this example, please see https://github.com/robbert-harms/MDT.


*******
Summary
*******
* Free software: LGPL v3 license
* Interface in Python, computations in OpenCL
* Implements Powell, Nelder-Mead Simplex and Levenberg-Marquardt non-linear optimization algorithms
* Implements various Markov Chain Monte Carlo (MCMC) sampling routines
* Tags: optimization, sampling, parallel, opencl, python


*****
Links
*****
* Full documentation: https://mot.readthedocs.org
* Project home: https://github.com/robbert-harms/MOT
* PyPi package: `PyPi <http://badge.fury.io/py/mot>`_


************************
Quick installation guide
************************
The basic requirements for MOT are:

* Python 3.x
* OpenCL 1.2 (or higher) support in GPU driver or CPU runtime


**Linux**

For Ubuntu >= 16 you can use:

* ``sudo add-apt-repository ppa:robbert-harms/cbclab``
* ``sudo apt update``
* ``sudo apt install python3-pip python3-mot``
* ``sudo pip3 install tatsu``


For Debian users and Ubuntu < 16 users, install MOT with:

* ``sudo apt install python3 python3-pip python3-pyopencl python3-devel``
* ``sudo pip3 install mot``


**Mac**

* Install Python Anaconda 3.* 64bit from https://www.continuum.io/downloads>
* Open a terminal and type ``pip install mot``


**Windows**
For Windows the short guide is:

* Install Python Anaconda 3.* 64bit from https://www.continuum.io/downloads
* Install or upgrade your GPU drivers
* Install PyOpenCL using one of the following methods:
    1. Use a binary, for example from http://www.lfd.uci.edu/~gohlke/pythonlibs/#pyopencl or;
    2. Compile PyOpenCL with ``pip install pyopencl``, this requires:
        * Visual Studio 2015 (Community edition or higher) with the Python and Common Tools for Visual C++ options enabled
        * OpenCL development kit (`NVidia CUDA <https://developer.nvidia.com/cuda-downloads>`_ or `Intel OpenCL SDK <https://software.intel.com/en-us/intel-opencl>`_ or the `AMD APP SDK <http://developer.amd.com/tools-and-sdks/opencl-zone/amd-accelerated-parallel-processing-app-sdk/>`_)
* Open a Anaconda shell and type: ``pip install mot``


For more information and for more elaborate installation instructions, please see: https://mot.readthedocs.org


*******
Caveats
*******
There are a few caveats and known issues, primarily related to OpenCL:

* Windows support is experimental due to the difficulty of installing PyOpenCL, hopefully installing PyOpenCL will get easier on Windows soon.
* GPU acceleration is not possible in most virtual machines due to lack of GPU or PCI-E pass-through, this will change whenever virtual machines vendors program this feature.
  Our recommendation is to install Linux on your machine directly.


CHANGELOG.rst
*********
Changelog
*********

v0.11.4 (2022-10-20)
====================

Added
-----
- Adds a few more Bessel functions.

Other
-----
- Removed the Ubuntu build for now.
- Upgraded the ubuntu distributions.
- Made MOT compatible with Python3.10 and the latest version of the TatSu library.


v0.11.3 (2020-05-10)
====================

Added
-----
- Adds a few more Bessel functions.


v0.11.2 (2020-03-26)
====================

Fixed
-----
- Fixed problems with device initialization in case the the context creation throws a runtime error.


v0.11.1 (2020-03-05)
====================

Fixed
-----
- Fixed pyopencl.LogicError: Context failed: invalid value happening when trying to create a context with a platform with no devices.

Other
-----
- Replaced maintainer e-mail address with personal e-mail address.


v0.11.0 (2020-02-14)
====================

Added
-----
- Adds an event system for better control of the processing.
- Adds a hashing method to the CLEnvironment.
- Adds use_host_ptr flag to the Array class to enable/disable that flag in the buffer allocation.
- Adds a flag to the Zeros kerneldata to enable non host accessible buffers (does not use_host_ptr)
- Adds inline keyword to the Scalar kernel data.
- Adds ctype property to the kernel data classes.
- Adds a do_data_transfer, is_blocking and wait_for flag to the evaluate method of a CLFunction.

Changed
-------
- Changed the get_subset signature to also accept a batch_range instead of batch_start and batch_end.
- Changed the get_subset method to check for consecutive problem indices and if so, use the batch range method instead.
- Cached the CLEnvironment creation to ensure same CLEnvironments have the same OpenCL queue.
- Removed the 'COAW' requirements from the array conversions, it was not needed.
- The CL function now caches the compiled kernels for faster consecutive execution.

Other
-----
- Set the ``enqueue_{host,device}_access`` methods of the KernelData to accept a list of CLEnvironments.
- Buffer allocation is now done per context instead of per CLEnvironment.


v0.10.0 (2020-01-25)
====================

Added
-----
- Adds kernel keyword to the opencl preprocessor.
- Adds the inverse-gamma distribution.
- Adds load balancing of CL compute tasks.
- Adds a buffer cache to the Array KernelData class

Changed
-------
- Updates the random123 library to version 1.13.2.
- Changed the execution to include an intermediate Processor object. This is meant for pipelining functions.
- Changed the kernel data array readout to include mapping the entire array if the array object did not specify parallelize_over_first_dimension.
- Changed the default random123 method to philox as it is slightly faster.

Fixed
-----
- Fixed bug in which data was modified twice towards mot_float_type.
- Small bugfix for when the last worker had no items left.
- Fixed that minimizer crashed if None data was given.

Other
-----
- Made the CLEnvironments share a context whenever possible.
- Created a new module for processors. Fixed the random number generation in the sampling.
- Moved the function processors out of the cl_function module.
- Moved the automatic wrapped kernel generation to a method of CLFunction.
- Moved the conversion of inputs to KernelData objects to a separate function.
- Allowed the configuration function set_cl_environments accept integers as arguments.
- CLRuntimeInfo now also supports initialization using integer indices and single value elements.


v0.9.6 (2019-12-17)
===================

Fixed
-----
- Fixed bug when dividing work over multiple devices. The batch size was provided incorrectly after changes in v0.9.5.

Other
-----
- Renamed the lib_ clfunctions to _spf functions. This formalizes the specialization a little bit better.
- Moved the Bracket method outside the Powell routine, making it a standalone routine now.


v0.9.5 (2019-12-16)
===================

Added
-----
- Adds switch to the minimize function to disable local reduction if so desired.
- Adds a utility function for easy splitting work over the work items.

Changed
-------
- Let the powell algorithm use local reduction.
- CL function will now process all provided work instances in one batch instead of splitting it in batches.

Other
-----
- Moved repository to private github account.


v0.9.4 (2019-12-06)
===================

Added
-----
- Added the method 'get_subset' to allow getting kernel data for a subset of the instances.

Changed
-------
- Changed the Array kernel data type by removing the offset_str in favor of a simple boolean switch.

Other
-----
- Bugfix in convert_data_to_dtype, in the case a vector type was already in the correct dtype.
- Improved error reporting in the function evaluator.


v0.9.3 (2019-06-04)
===================

Added
-----
- Adds include guards to the Struct kernel data object.


v0.9.2 (2019-04-01)
===================

Added
-----
- Makes CL functions directly callable. Calls are forwarded to the evaluate method.


v0.9.1 (2019-02-21)
===================

Added
-----
- Adds OpenCL approximations to the Gamma CDF and Gamma PPF functions.

Other
-----
- Removed atomic functions. They were unsupport on some platforms.
- Made the ML routines more robust for objective functions that return NaN.


v0.9.0 (2019-02-19)
===================
The primary addition in this version is the support for inequality constraints in the non-linear optimization routines.
This allows the user to provide a function :math:`g(x)` enforcing the rule :math:`g(x) <= 0`. The current optimization routines
have no native capability for dealing with these bounds and therefore use the penalty method.

Added
-----
- Adds support for inequality constraints (in the non-linear optimization routines) using the penalty method.
- The CL function parser can now handle ``/* */`` documentation strings prefixed to the function definition.
- Adds atomic float and double functions to the CL kernels

Changed
-------
- Moved the box constraints to the penalty methods.

pppe/tools.py
from collections import defaultdict
import numbers
import pyopencl.array as cl_array
import numpy as np
import re
import pyopencl as cl
import scipy.special
from functools import reduce

__author__ = 'Robbert Harms'
__date__ = "2014-05-13"
__license__ = "LGPL v3"
__maintainer__ = "Robbert Harms"
__email__ = "robbert.harms@maastrichtuniversity.nl"


def device_supports_double(cl_device):
    """Check if the given CL device supports double

    Args:
        cl_device (pyopencl cl device): The device to check if it supports double.

    Returns:
        True if the given cl_device supports double, false otherwise.
    """
    return cl_device.get_info(cl.device_info.DOUBLE_FP_CONFIG) == 63


def results_to_dict(results, param_names):
    """Create a dictionary out of the results, which the optimizer can output to the user

    Args:
        results: a 2d (from optimization) or 3d (from sampling) array that needs to be converted to a dictionary.
        param_names (list of str): the names of the parameters, one per column

    Returns:
        dict: the results packed in a dictionary
    """
    s = results.shape
    d = {}
    if len(s) == 2:
        for i in range(len(param_names)):
            d[param_names[i]] = results[:, i]
    else:
        for i in range(len(param_names)):
            d[param_names[i]] = results[:, i, :]
    return d


def set_cl_compatible_data_type(value, data_type):
    """Set the given value (numpy array) to the given data type, one which is CL compatible.

    Args:
        - value: numpy array
        - data type: 'float', 'double' or 'double<int>' or 'float<int>'

    Returns:
        The same array, but then with the correct data type. If the data type includes an integer the value is parsed
        to a CL vector type.
    """
    is_vector = bool(re.compile('\d').search(data_type))
    if is_vector:
        return array_to_cl_vector(value, data_type)
    else:
        s = data_type.replace('*', '')
        s = s.replace(' ', '')

        if s == 'float':
            if isinstance(value, numbers.Number):
                return np.float32(value)
            return value.astype(np.float32)
        elif s == 'int':
            if isinstance(value, numbers.Number):
                return np.int32(value)
            return value.astype(np.int32)
        else:
            if isinstance(value, numbers.Number):
                return np.float64(value)
            return value.astype(np.float64)


def numpy_types_to_cl(data_type):
    """Get the CL type name of the given numpy type. Call this function with argument data.dtype.type."""
    names = {np.float32: 'float',
             np.float64: 'double',
             np.int32: 'int',
             np.int64: 'long',
             np.uint32: 'uint',
             np.uint64: 'ulong'}
    return names[data_type]


def get_opencl_vector_data_type(vector_length, data_type):
    """Get the data type for a vector of the given length and given type.

    Args:
        - vector_length: the length of the CL vector data type
        - data_type: the data/double type of the data
    Returns:
        The vector type given the given vector length and data type
    """
    if vector_length not in (2, 3, 4, 8, 16):
        raise ValueError('The given vector length is not one of (2, 3, 4, 8, 16)')
    if data_type not in ('char', 'uchar', 'short', 'ushort', 'int', 'uint', 'long', 'ulong', 'float', 'double', 'half'):
        raise ValueError('The given data type ({}) is not supported.'.format(data_type))

    return getattr(cl_array.vec, data_type + repr(vector_length))


def array_to_cl_vector(array, data_type, vector_length=None):
    """Create a CL vector type of the given array.

    Standard is chosen for the minimum vector length that can hold the given array, else, if vector_length is specified
    the given length is used, but only if vector length is one of (2, 3, 4, 8, 16).

    Args:
        array: the array of which to translate each row to a vector
        vector_length: if specified (non-None) the desired vector length. It must be one of (2, 3, 4, 8, 16)

    Returns:
        An array of the same length as the given array, but with only one column per row. This column contains the
        opencl vector.

    Raises:
        ValueError: if the vector length is not one of (2, 3, 4, 8, 16)
    """
    s = array.shape
    if len(s) > 1:
        width = s[1]
    else:
        width = 1

    if vector_length is None:
        vector_length = width

    if 'double' in data_type:
        dtype = get_opencl_vector_data_type(vector_length, 'double')
    else:
        dtype = get_opencl_vector_data_type(vector_length, 'float')

    ve = np.zeros((s[0], 1), dtype=dtype, order='C')
    for i in range(s[0]):
        for j in range(width):
            ve[i, 0][j] = array[i, j]

    return ve


def is_cl_vector_type(data):
    """Check if the given numpy array contains a CL vector data type"""
    return len(data.dtype.base.descr) > 1


def vector_type_lookup(data):
    """Return the correct vector type, vector length and type_name ('float' or 'double')

    Args:
        - data: a numpy vector or array of which we want to know the CL data type.

    Returns:
        A dictionary with the keys
            - dtype: the cl data type as a object
            - length: the length of the vector
            - data_type: the type of data, double, float etc.
            - cl_name: <data_type><length>, i.e. the full name of the type for use in CL code

        Returns None if it is an unknown type.
    """
    length = len(data.dtype.base.descr)
    if length < 2:
        return None
    cl_type_name = _numpy_to_cl_dtype_names(data.dtype.fields['x'][0])
    return {'dtype': data.dtype, 'length': length, 'data_type': cl_type_name, 'cl_name': (cl_type_name + repr(length))}


def _numpy_to_cl_dtype_names(cl_data_type_name):
    """Translates the names from numpy types to CL types.

    Example: float64 -> double
    """
    if isinstance(cl_data_type_name, np.dtype):
        cl_data_type_name = cl_data_type_name.name

    m = {'float64': 'double',
         'float32': 'float'}
    if cl_data_type_name in m:
        return m[cl_data_type_name]
    return None


def get_cl_double_extension_definer(platform):
    return '''
        #if defined(cl_khr_fp64)
        #pragma OPENCL EXTENSION cl_khr_fp64 : enable
        #elif defined(cl_amd_fp64)
        #pragma OPENCL EXTENSION cl_amd_fp64 : enable
        #endif
    '''


def get_bessel_roots(number_of_roots=30):
    """These roots are used in some of the compartment models. It are the roots of the equation J'_1(x) = 0.

    That is, where J_1 is the first order Bessel function of the first kind.
    """
    return scipy.special.jnp_zeros(1, number_of_roots).astype(np.float64)


class TopologicalSort(object):

    def __init__(self, data):
        self.data = data

    def get_sorted(self):
        """Topological sort the the given data. The data should consist of a dictionary structure.

        Args:
            data (dict); dictionary structure where the value is a list of dependencies for that given key.
                Example: {'a', (), 'b': ('a',)}, here a depends on nothing and b depends on a.
        """
        if not len(self.data):
            return

        new_data = {}
        for k, v in self.data.items():
            new_v = set([])
            if v:
                new_v = set([e for e in v if e is not k])
            new_data.update({k: new_v})
        data = new_data

        # Find all items that don't depend on anything.
        extra_items_in_deps = reduce(set.union, data.values()) - set(data.keys())

        # Add empty dependences where needed.
        data.update({item: set() for item in extra_items_in_deps})
        while True:
            ordered = set(item for item, dep in data.items() if len(dep) == 0)
            if not ordered:
                break
            yield ordered
            data = {item: (dep - ordered) for item, dep in data.items() if item not in ordered}

        if len(data) != 0:
            raise ValueError('Cyclic dependencies exist among these items: {}'.format(', '.join(repr(x)
                                                                                                for x in data.items())))

    def get_flattened_sort(self, sort=True):
        """Returns a single list of dependencies. For any set returned by
            toposort(), those items are sorted and appended to the result (just to
            make the results deterministic).
        """
        result = []
        for d in self.get_sorted():
            result.extend((sorted if sort else list)(d))
        return result


def get_read_only_cl_mem_flags(cl_environment):
    """Get the right read only flags to use in the given environment."""
    if cl_environment.is_gpu:
        return cl.mem_flags.READ_ONLY | cl.mem_flags.COPY_HOST_PTR
    else:
        return cl.mem_flags.READ_ONLY | cl.mem_flags.USE_HOST_PTR


def get_read_write_cl_mem_flags(cl_environment):
    """Get the right write only flags to use in the given environment."""
    if cl_environment.is_gpu:
        return cl.mem_flags.READ_WRITE | cl.mem_flags.COPY_HOST_PTR
    else:
        return cl.mem_flags.READ_WRITE | cl.mem_flags.USE_HOST_PTR


def get_write_only_cl_mem_flags(cl_environment):
    """Get the right write only flags to use in the given environment."""
    if cl_environment.is_gpu:
        return cl.mem_flags.WRITE_ONLY | cl.mem_flags.COPY_HOST_PTR
    else:
        return cl.mem_flags.WRITE_ONLY | cl.mem_flags.USE_HOST_PTR


def check_array_fits_cl_memory(array, dtype, max_size):
    """Check if the given array when casted to the given type can be fit into the given max_size"""
    return np.product(array.shape) * np.dtype(dtype).itemsize < max_size


def get_cl_data_type_from_data(data):
    """Get the data attributes for the given dataset, this can be used for building a CL kernel"""
    if is_cl_vector_type(data):
        vector_info = vector_type_lookup(data)
        length = vector_info['length']
        if vector_info['length'] % 2 == 1:
            length += 1
        return vector_info['data_type'] + repr(length)
    return numpy_types_to_cl(data.dtype.type)


def get_correct_cl_data_type_from_data(data):
    """Get the data type for the given data"""
    if is_cl_vector_type(data):
        vector_info = vector_type_lookup(data)
        length = vector_info['length']
        if length % 2 == 1:
            length += 1
        return get_opencl_vector_data_type(length, data_type=vector_info['data_type'])
    return data.dtype.type


def set_correct_cl_data_type(data):
    """Set for the given data the data type given by get_correct_cl_data_type_from_data()"""
    if data is not None:
        if isinstance(data, dict):
            for key, d in data.items():
                data[key] = set_correct_cl_data_type(d)
            return data
        elif isinstance(data, (tuple, list)):
            items = []
            for d in data:
                items.append(set_correct_cl_data_type(d))
            return items
        elif isinstance(data, (numbers.Number,)):
            return np.array(data, dtype=np.float64, order='C')
        else:
            return data.astype(get_correct_cl_data_type_from_data(data))
    return None


class ParameterCLCodeGenerator(object):

    def __init__(self, device, var_data_dict, prtcl_data_dict, model_data_dict, add_var_data_multipliers=True):
        self._device = device
        self._max_constant_buffer_size = device.get_info(cl.device_info.MAX_CONSTANT_BUFFER_SIZE)
        self._max_constant_args = device.get_info(cl.device_info.MAX_CONSTANT_ARGS)
        self._var_data_dict = var_data_dict
        self._prtcl_data_dict = prtcl_data_dict
        self._model_data_dict = model_data_dict
        self._add_var_data_multipliers = add_var_data_multipliers
        self._kernel_items = self._get_all_kernel_source_items()

    def get_data_struct(self):
        return self._kernel_items['data_struct']

    def get_kernel_param_names(self):
        return self._kernel_items['kernel_param_names']

    def get_data_struct_init_assignment(self, variable_name):
        return 'optimize_data ' + variable_name + ' = {' + ', '.join(self._kernel_items['data_struct_init']) + '};'

    def _get_all_kernel_source_items(self):
        """Get the CL strings for the kernel source items for most common CL kernels in this library."""
        constant_args_counter = 0

        kernel_param_names = []
        data_struct_init = []
        data_struct_names = []

        for key, vdata in self._var_data_dict.items():
            clmemtype = 'global'

            if check_array_fits_cl_memory(vdata, get_correct_cl_data_type_from_data(vdata),
                                          self._max_constant_buffer_size):
                if constant_args_counter < self._max_constant_args:
                    clmemtype = 'constant'
                    constant_args_counter += 1

            param_name = 'var_data_' + key
            cl_data_type = get_cl_data_type_from_data(vdata)

            kernel_param_names.append(clmemtype + ' ' + cl_data_type + '* ' + param_name)
            data_struct_names.append(clmemtype + ' ' + cl_data_type + '* ' + param_name)

            if self._add_var_data_multipliers:
                mult = vdata.shape[1] if len(vdata.shape) > 1 else 1
                data_struct_init.append(param_name + ' + gid * ' + repr(mult))
            else:
                data_struct_init.append(param_name)

        for key, vdata in self._prtcl_data_dict.items():
            clmemtype = 'global'

            if check_array_fits_cl_memory(vdata, get_correct_cl_data_type_from_data(vdata),
                                          self._max_constant_buffer_size):
                if constant_args_counter < self._max_constant_args:
                    clmemtype = 'constant'
                    constant_args_counter += 1

            param_name = 'prtcl_data_' + key
            cl_data_type = get_cl_data_type_from_data(vdata)

            kernel_param_names.append(clmemtype + ' ' + cl_data_type + '* ' + param_name)
            data_struct_init.append(param_name)
            data_struct_names.append(clmemtype + ' ' + cl_data_type + '* ' + param_name)

        for key, vdata in self._model_data_dict.items():
            clmemtype = 'global'
            param_name = 'fixed_data_' + key
            cl_data_type = get_cl_data_type_from_data(vdata)

            data_struct_init.append(param_name)

            if isinstance(vdata, np.ndarray):
                kernel_param_names.append(clmemtype + ' ' + cl_data_type + '* ' + param_name)
                data_struct_names.append(clmemtype + ' ' + cl_data_type + '* ' + param_name)
            else:
                kernel_param_names.append(cl_data_type + ' ' + param_name)
                data_struct_names.append(cl_data_type + ' ' + param_name)

        data_struct = '''
            typedef struct{
                ''' + ('' if data_struct_names else 'constant void* place_holder;') + '''
                ''' + " ".join((name + ";\n" for name in data_struct_names)) + '''
            } optimize_data;
        '''

        return {'kernel_param_names': kernel_param_names,
                'data_struct_names': data_struct_names,
                'data_struct_init': data_struct_init,
                'data_struct': data_struct}


def init_dict_tree():
    return defaultdict(init_dict_tree)